glue/src/main/scala/com/gu/tableversions/glue/GlueMetastore.scala

```diff
@@ -0,0 +1,184 @@
+      else
+        getPartitionedTableVersion(table, tableLocation, partitionColumns)
+    } yield (version)
+  }
+
+  private def getPartitionedTableVersion(
+      table: TableName,
+      tableLocation: URI,
+      partitionColumns: List[PartitionColumn]): F[TableVersion] = {
+
+    def toPartition(columnValues: List[ColumnValue]): Partition = columnValues match {
+      case head :: tail => Partition(head, tail: _*)
+      case _            => throw new Exception("empty columnValues list for partition found")
+    }
+
+    def toPartitionWithVersion(gluePartition: GluePartition): (Partition, Version) = {
+
+      val partitionColumnAndValue: List[(PartitionColumn, String)] = partitionColumns.zip(gluePartition.getValues)
+      val columnValues: List[ColumnValue] = partitionColumnAndValue.map { case (key, value) => ColumnValue(key, value) }
+      val partition: Partition = toPartition(columnValues)
+      val location = new URI(gluePartition.getStorageDescriptor.getLocation)
+      partition -> VersionPaths.parseVersion(location)
+    }
+
+    getGluePartitions(table).map { gluePartitions =>
+      val partitionVersions: Map[Partition, Version] = gluePartitions.map(toPartitionWithVersion).toMap
+      PartitionedTableVersion(partitionVersions)
+    }
+
+  }
+
+  override def update(table: TableName, changes: Metastore.TableChanges): F[Unit] =
+    changes.operations.traverse_(appliedOp(table))
+
+  private def appliedOp(table: TableName)(operation: TableOperation): F[Unit] =
+    operation match {
+      case AddPartition(partition, version)           => addPartition(table, partition, version)
+      case UpdatePartitionVersion(partition, version) => updatePartitionVersion(table, partition, version)
+      case RemovePartition(partition)                 => removePartition(table, partition)
+      case UpdateTableVersion(versionNumber)          => updateTableLocation(table, versionNumber)
+    }
+
+  def addPartition(table: TableName, partition: Partition, version: Version): F[Unit] = {
+
+    def partitionLocation(tableLocation: URI): String = {
+      val unversionedLocation: String = partition.resolvePath(tableLocation).toString
+      if (version == Version.Unversioned) unversionedLocation
+      else
+        unversionedLocation + version.label
+    }
+    for {
+      glueTable <- getGlueTable(table)
+      tableLocation <- F.delay(findTableLocation(glueTable))
+      location = partitionLocation(tableLocation)
+      storageDescriptor <- F.delay(new StorageDescriptor().withLocation(location))
+      partitionValues <- F.delay(partition.columnValues.map(_.value).toList)
+      input <- F.delay(new PartitionInput().withValues(partitionValues).withStorageDescriptor(storageDescriptor))
+
+      addPartitionRequest = new CreatePartitionRequest()
+        .withDatabaseName(table.schema)
+        .withTableName(table.name)
+        .withPartitionInput(input)
+
+      _ <- F.delay { glue.createPartition(addPartitionRequest) }
```

Comment:
Maybe worth pulling out an `addPartition` method to match the other cases, e.g. `removePartition` etc? Whatever you prefer anyway.

Reply:
(Actually, I misread the rest of the code, so just ignore the above comment!)